### PR TITLE
doc-improvement: applications: nrf_desktop: Document USB next support

### PR DESCRIPTION
**Pull request by MarekPieta**
Change adds USB next stack integration documentation for USB state application module.

Jir: NCSDK-27661

### CODE REVIEW

File: applications/nrf_desktop/doc/usb_state.rst [hunk 1]
```diff
@@ -24,60 +24,67 @@ Configuration
...
-=================================
-
-The nRF Desktop device can provide multiple instances of a HID-class USB device.
-The number of instances is controlled by :kconfig:option:`CONFIG_USB_HID_DEVICE_COUNT`.
-By default, the option is set to:
+The nRF Desktop device can provide multiple instances of HID-class USB.
+By default, the number of the HID-class USB instances should be set to:
 
 * ``1`` for HID peripherals (:ref:`CONFIG_DESKTOP_ROLE_HID_PERIPHERAL <config_desktop_app_options>`).
+  More HID-class USB instances can be used if selective HID report subscription feature is enabled (:ref:`CONFIG_DESKTOP_USB_SELECTIVE_REPORT_SUBSCRIPTION <config_desktop_app_options>`).
```
Comment (divipillai) [inline applications/nrf_desktop/doc/usb_state.rst:74]:
```suggestion
  You can use more HID-class USB instances if the selective HID report subscription feature is enabled (:ref:`CONFIG_DESKTOP_USB_SELECTIVE_REPORT_SUBSCRIPTION <config_desktop_app_options>`).
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 2]
```diff
@@ -24,60 +24,67 @@ Configuration
...
 ----------------------
 
-The nRF Desktop Peripheral devices by default use only a single HID-class USB instance.
+The nRF Desktop peripheral devices by default use only a single HID-class USB instance.
 In that case, this instance is used for all the HID reports.
 
-Enable :ref:`CONFIG_DESKTOP_USB_SELECTIVE_REPORT_SUBSCRIPTION <config_desktop_app_options>` to use more than one HID-class USB instance on nRF Desktop Peripheral.
-Make sure to set a greater value in the :kconfig:option:`CONFIG_USB_HID_DEVICE_COUNT` option and create an additional :file:`usb_state_def.h` header in the configuration.
-The header assigns HID reports to the HID-class USB instances.
+Enable :ref:`CONFIG_DESKTOP_USB_SELECTIVE_REPORT_SUBSCRIPTION <config_desktop_app_options>` to use more than one USB HID-class instance on nRF Desktop peripheral.
```
Comment (divipillai) [inline applications/nrf_desktop/doc/usb_state.rst:85]:
```suggestion
Enable :ref:`CONFIG_DESKTOP_USB_SELECTIVE_REPORT_SUBSCRIPTION <config_desktop_app_options>` to use more than one USB HID-class instance on the nRF Desktop peripheral.
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 3]
```diff
@@ -102,23 +109,56 @@ For example, the file contents can look as follows:
...
+If only one instance is used, reports from all peripherals connected to the dongle are forwarded to the same instance.
 In other cases, reports from each of the bonded peripherals are forwarded to a dedicated HID-class USB instance.
 The same instance is used after reconnection.
 
+USB HID configuration in USB legacy stack
+-----------------------------------------
+
+For the USB legacy stack, the :ref:`CONFIG_DESKTOP_USB_STACK_LEGACY <config_desktop_app_options>` selects the required dependencies: :kconfig:option:`CONFIG_USB_DEVICE_STACK` and :kconfig:option:`CONFIG_USB_DEVICE_HID` Kconfig options.
+The number of HID-class USB instances in the USB legacy stack is specified by :kconfig:option:`CONFIG_USB_HID_DEVICE_COUNT`.
+The default value of this option is updated to match requirements for either nRF Desktop peripheral or dongle.
```
Comment (divipillai) [inline applications/nrf_desktop/doc/usb_state.rst:125]:
```suggestion
The default value of this option is updated to match requirements for either an nRF Desktop peripheral or a dongle.
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 4]
```diff
@@ -102,23 +109,56 @@ For example, the file contents can look as follows:
...
+The module aligns the USB HID boot protocol configuration with the application configuration.
+The module sets the default value of the :kconfig:option:`CONFIG_USB_HID_BOOT_PROTOCOL` Kconfig option and calls the :c:func:`usb_hid_set_proto_code` function during initialization to set the USB HID boot interface protocol code.
+Common HID boot protocol code is used for all of the HID-class USB instances.
+
+USB HID configuration in USB next stack
+---------------------------------------
+
+For the USB next stack, the :ref:`CONFIG_DESKTOP_USB_STACK_NEXT <config_desktop_app_options>` selects the :kconfig:option:`CONFIG_USB_DEVICE_STACK_NEXT` Kconfig option.
+Every USB HID-class instance is configured through a separate DTS node compatible with ``zephyr,hid-device``.
+The DTS node configures, among others, used HID boot protocol, size of the longest HID input report and HID polling rate.
```
Comment (divipillai) [inline applications/nrf_desktop/doc/usb_state.rst:146]:
```suggestion
The DTS node configures, among others, used the HID boot protocol, the size of the longest HID input report, and the HID polling rate.
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 5]
```diff
@@ -102,23 +109,56 @@ For example, the file contents can look as follows:
...
+Make sure to update DTS configuration to match requirements of your application.
```
Comment (divipillai) [inline applications/nrf_desktop/doc/usb_state.rst:147]:
```suggestion
Make sure to update the DTS configuration to match requirements of your application.
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 6]
```diff
@@ -24,60 +24,67 @@ Configuration
 *************
 
 The module is enabled by selecting :ref:`CONFIG_DESKTOP_USB_ENABLE <config_desktop_app_options>` option.
```
Comment (annwoj) [inline applications/nrf_desktop/doc/usb_state.rst:26]:
```suggestion
The module is enabled by selecting the :ref:`CONFIG_DESKTOP_USB_ENABLE <config_desktop_app_options>` option.
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 7]
```diff
@@ -24,60 +24,67 @@ Configuration
...
+
+* :ref:`CONFIG_DESKTOP_USB_STACK_LEGACY <config_desktop_app_options>`
+  The option enables USB legacy stack.
+* :ref:`CONFIG_DESKTOP_USB_STACK_NEXT <config_desktop_app_options>`
+  The option enables USB next stack.
+
+.. note::
+  The USB next stack integration is ref:`experimental <software_maturity>`.
+  The HID boot protocol integration is not yet fully tested and may not work properly.
+  The secondary image slot background erase in :ref:`nrf_desktop_dfu` may cause missing USB HID subscriptions after an USB cable is attached.
```
Comment (annwoj) [inline applications/nrf_desktop/doc/usb_state.rst:43]:
```suggestion
  The secondary image slot background erase in :ref:`nrf_desktop_dfu` may cause missing USB HID subscriptions after a USB cable is attached.
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 8]
```diff
@@ -102,23 +109,56 @@ For example, the file contents can look as follows:
...
+-----------------------------------------
+
+For the USB legacy stack, the :ref:`CONFIG_DESKTOP_USB_STACK_LEGACY <config_desktop_app_options>` selects the required dependencies: :kconfig:option:`CONFIG_USB_DEVICE_STACK` and :kconfig:option:`CONFIG_USB_DEVICE_HID` Kconfig options.
+The number of HID-class USB instances in the USB legacy stack is specified by :kconfig:option:`CONFIG_USB_HID_DEVICE_COUNT`.
+The default value of this option is updated to match requirements for either nRF Desktop peripheral or dongle.
+
+Low latency device configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The module sets default value of :kconfig:option:`CONFIG_USB_HID_POLL_INTERVAL_MS` to ``1``.
```
Comment (annwoj) [inline applications/nrf_desktop/doc/usb_state.rst:130]:
```suggestion
The module sets default the value of :kconfig:option:`CONFIG_USB_HID_POLL_INTERVAL_MS` to ``1``.
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 9]
```diff
@@ -292,6 +292,7 @@ nRF Desktop
   * A warning log for handling ``-EACCES`` error code returned by functions that send GATT notification with HID report in :ref:`nrf_desktop_hids`.
     The error code might be returned if an HID report is sent right after a remote peer unsubscribes.
     The warning log prevents displaying an error log in a use case that does not indicate an error.
+  * Experimental support for USB next stack (:kconfig:option:`CONFIG_USB_DEVICE_STACK_NEXT`) to :ref:`nrf_desktop_usb_state`.
```
Comment (annwoj) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:295]:
```suggestion
  * Experimental support for the USB next stack (:kconfig:option:`CONFIG_USB_DEVICE_STACK_NEXT`) to :ref:`nrf_desktop_usb_state`.
```

File: applications/nrf_desktop/doc/usb_state.rst [hunk 10]
```diff
@@ -137,17 +180,22 @@ From the application's viewpoint, USB can be in the following states:
...
+These states are broadcasted by the |usb_state| with a :c:struct:`usb_state_event`.
 When the device is connected to the host and configured for the communication, the module will broadcast the :c:enum:`USB_STATE_ACTIVE` state.
-The module will also subscribe to all HID reports available in the application for the selected protocol.
 
-When the device is disconnected from the host, the module will unsubscribe from receiving the HID reports.
+HID data handling
+=================
+
+The module handles USB HID input report subscriptions.
+For the USB legacy stack, all of the USB HID-class instances subscribe to all supported HID reports for the selected protocol when USB stack reports :c:enum:`USB_DC_CONFIGURED` and :c:enum:`USB_STATE_ACTIVE` state is broadcasted.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/nrf_desktop/doc/usb_state.rst:190]: Please double check if it shouldn't be `is` here. `:c:enum:`USB_STATE_ACTIVE` state is broadcasted` (just one thing). The `USB_DC_CONFIGURED` is reported by the USB stack.

File: applications/nrf_desktop/doc/usb_state.rst [hunk 11]
```diff
@@ -23,61 +23,68 @@ Module events
...
-The option selects :kconfig:option:`CONFIG_USB_DEVICE_STACK` and :kconfig:option:`CONFIG_USB_DEVICE_HID` Kconfig options.
+The module is enabled by selecting the :ref:`CONFIG_DESKTOP_USB_ENABLE <config_desktop_app_options>` option.
 
 The :ref:`CONFIG_DESKTOP_USB_ENABLE <config_desktop_app_options>` option is implied by the :ref:`CONFIG_DESKTOP_ROLE_HID_DONGLE <config_desktop_app_options>` option.
 The module is enabled by default for the nRF Desktop dongles because the dongles forward the HID data to the host connected over USB.
 See the :ref:`nrf_desktop_hid_configuration` documentation for details.
 
-Additionally, you can also configure the options described in the following sections.
+The module can be used with either the USB legacy stack or the USB next stack.
+The USB stack is selected by enabling one of the following Kconfig options:
```
Comment (zycz) [inline applications/nrf_desktop/doc/usb_state.rst:33]: "choice option" - to stress that you cannot select both?

File: applications/nrf_desktop/doc/usb_state.rst [hunk 12]
```diff
@@ -23,61 +23,68 @@ Module events
...
+
+* :ref:`CONFIG_DESKTOP_USB_STACK_LEGACY <config_desktop_app_options>`
+  The option enables USB legacy stack.
```
Comment (zycz) [inline applications/nrf_desktop/doc/usb_state.rst:36]: Maybe mention that it is default option?

File: applications/nrf_desktop/doc/usb_state.rst [hunk 13]
```diff
@@ -23,61 +23,68 @@ Module events
...
+* :ref:`CONFIG_DESKTOP_USB_STACK_NEXT <config_desktop_app_options>`
+  The option enables USB next stack.
+
+.. note::
+  The USB next stack integration is ref:`experimental <software_maturity>`.
+  The HID boot protocol integration is not yet fully tested and may not work properly.
+  The secondary image slot background erase in :ref:`nrf_desktop_dfu` may cause missing USB HID subscriptions after a USB cable is attached.
+
+Some of the properties are configured in the same way for both stacks, but part of the configuration is specific to the selected USB stack.
+See the following section for details.
```
Comment (zycz) [inline applications/nrf_desktop/doc/usb_state.rst:46]: section -> sections

File: applications/nrf_desktop/doc/usb_state.rst [hunk 14]
```diff
@@ -102,23 +109,56 @@ For example, the file contents can look as follows:
...
+
+Defining USB HID-class instances in the DTS, by default, enables the :kconfig:option:`CONFIG_USBD_HID_SUPPORT` Kconfig option.
+
 USB wakeup configuration
 ========================
 
 The nRF Desktop device can work as a source of wakeup events for the host device if connected through the USB.
 To use the feature, enable the :ref:`CONFIG_DESKTOP_USB_REMOTE_WAKEUP <config_desktop_app_options>` option.
-This option selects the :kconfig:option:`CONFIG_USB_DEVICE_REMOTE_WAKEUP` to enable required dependencies in the USB stack.
+This option selects the :kconfig:option:`CONFIG_USB_DEVICE_REMOTE_WAKEUP` Kconfig option to enable required dependencies if USB legacy stack is used.
```
Comment (zycz) [inline applications/nrf_desktop/doc/usb_state.rst:158]: Shouldn't we mention what happens in next-stack here?

**Review by zycz — COMMENTED** [review]
General remark - I find it a bit confusing to what stack (both/legacy/next) each part is referring - maybe there is a way to better distinguish between them?

File: applications/nrf_desktop/doc/usb_state.rst (hunk 14 above)
Comment (MarekPieta) [reply applications/nrf_desktop/doc/usb_state.rst:158]: No additional configuration is applied there

File: applications/nrf_desktop/doc/usb_state.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply applications/nrf_desktop/doc/usb_state.rst:191]: yeah agree with Anna here.

File: applications/nrf_desktop/doc/usb_state.rst [hunk 15]
```diff
@@ -23,61 +23,69 @@ Module events
...
+The USB stack is selected by enabling one of the following Kconfig choice options:
+
+* :ref:`CONFIG_DESKTOP_USB_STACK_LEGACY <config_desktop_app_options>`
+  The option enables USB legacy stack (:kconfig:option:`CONFIG_USB_DEVICE_STACK`).
+  The USB legacy stack is used by default.
+* :ref:`CONFIG_DESKTOP_USB_STACK_NEXT <config_desktop_app_options>`
+  The option enables USB next stack (:kconfig:option:`CONFIG_USB_DEVICE_STACK_NEXT`).
+
+.. note::
+  The USB next stack integration is ref:`experimental <software_maturity>`.
```
Comment (divipillai) [inline applications/nrf_desktop/doc/usb_state.rst:42]:
```suggestion
  The USB next stack integration is :ref:`experimental <software_maturity>`.
```

**Review by divipillai — APPROVED** [review]

File: applications/nrf_desktop/doc/usb_state.rst (hunk 14 above)
Comment (zycz) [reply applications/nrf_desktop/doc/usb_state.rst:159]: But just reading this doc I have no idea is wakeup events working for next-stack. I would clearly state it here.

**Review by annwoj — APPROVED** [review]

File: applications/nrf_desktop/doc/usb_state.rst (hunk 14 above)
Comment (MarekPieta) [reply applications/nrf_desktop/doc/usb_state.rst:159]: Ok, added one more sentence. Hope it helps

Comment (zycz) [reply applications/nrf_desktop/doc/usb_state.rst:159]: I meant to add sentence that states that wakeup events are working on next-stack:
If the USB next stack is used, no additional Kconfig dependencies need to be selected. -> 
If the USB next stack is used,  wakeup events are enabled by default and do not need any option to turn them on.

Comment (MarekPieta) [reply applications/nrf_desktop/doc/usb_state.rst:159]: Discussed offline. Rephrased a bit.

**Review by zycz — APPROVED** [review]

File: applications/nrf_desktop/doc/usb_state.rst [hunk 16]
```diff
@@ -102,23 +110,57 @@ For example, the file contents can look as follows:
...
+Defining USB HID-class instances in the DTS, by default, enables the :kconfig:option:`CONFIG_USBD_HID_SUPPORT` Kconfig option.
+
 USB wakeup configuration
 ========================
 
 The nRF Desktop device can work as a source of wakeup events for the host device if connected through the USB.
 To use the feature, enable the :ref:`CONFIG_DESKTOP_USB_REMOTE_WAKEUP <config_desktop_app_options>` option.
-This option selects the :kconfig:option:`CONFIG_USB_DEVICE_REMOTE_WAKEUP` to enable required dependencies in the USB stack.
+This option selects the :kconfig:option:`CONFIG_USB_DEVICE_REMOTE_WAKEUP` Kconfig option to enable required dependencies if USB legacy stack is used.
+The USB next stack supports remote wakeup by default, and you need not select any additional Kconfig dependencies.
```
Comment (alstrzebonski) [inline applications/nrf_desktop/doc/usb_state.rst:160]: `don't need to select`?

Comment (MarekPieta) [reply applications/nrf_desktop/doc/usb_state.rst:160]: I think we use `need not select` in similar places. @divipillai could you confirm, please?

**Review by alstrzebonski — APPROVED** [review]

File: applications/nrf_desktop/doc/usb_state.rst (hunk 16 above)
Comment (MarekPieta) [reply applications/nrf_desktop/doc/usb_state.rst]: Discussed offline and aligned

**Pull request merged**